Subject: [PATCH] Implement startup health checks for Menu and Basket services to ensure readiness before traffic.
---
Index: webapp/src/main/kotlin/io/ktor/foodies/server/WebApp.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/webapp/src/main/kotlin/io/ktor/foodies/server/WebApp.kt b/webapp/src/main/kotlin/io/ktor/foodies/server/WebApp.kt
--- a/webapp/src/main/kotlin/io/ktor/foodies/server/WebApp.kt	(revision c368a9d202cfb2420ff9893ed9e4d90a7c6e0d87)
+++ b/webapp/src/main/kotlin/io/ktor/foodies/server/WebApp.kt	(date 1768469681069)
@@ -30,7 +30,7 @@
     install(Cohort) {
         verboseHealthCheckResponse = true
 
-        healthcheck("/healthz/startup", HealthCheckRegistry(Dispatchers.Default))
+        healthcheck("/healthz/startup", module.startupCheck)
         healthcheck("/healthz/liveness", HealthCheckRegistry(Dispatchers.Default))
         healthcheck("/healthz/readiness", module.readinessCheck)
     }
Index: webapp/src/main/kotlin/io/ktor/foodies/server/WebAppModule.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/webapp/src/main/kotlin/io/ktor/foodies/server/WebAppModule.kt b/webapp/src/main/kotlin/io/ktor/foodies/server/WebAppModule.kt
--- a/webapp/src/main/kotlin/io/ktor/foodies/server/WebAppModule.kt	(revision c368a9d202cfb2420ff9893ed9e4d90a7c6e0d87)
+++ b/webapp/src/main/kotlin/io/ktor/foodies/server/WebAppModule.kt	(date 1768468579077)
@@ -2,6 +2,7 @@
 
 import com.sksamuel.cohort.HealthCheckRegistry
 import com.sksamuel.cohort.healthcheck.http.EndpointHealthCheck
+import com.sksamuel.cohort.healthcheck.http.EndpointStartupHealthCheck
 import com.sksamuel.cohort.lettuce.RedisHealthCheck
 import io.ktor.client.HttpClient
 import io.ktor.client.engine.apache5.Apache5
@@ -12,6 +13,7 @@
 import io.ktor.foodies.server.htmx.menu.HttpMenuService
 import io.ktor.foodies.server.htmx.menu.MenuService
 import io.ktor.foodies.server.security.RedisSessionStorage
+import io.ktor.http.isSuccess
 import io.ktor.serialization.kotlinx.json.json
 import io.ktor.server.application.Application
 import io.ktor.server.application.ApplicationStopped
@@ -25,6 +27,7 @@
     val menuService: MenuService,
     val basketService: BasketService,
     val httpClient: HttpClient,
+    val startupCheck: HealthCheckRegistry,
     val readinessCheck: HealthCheckRegistry,
     val sessionStorage: SessionStorage
 )
@@ -47,6 +50,16 @@
         client.shutdown()
     }
 
+    val startupCheck = HealthCheckRegistry(Dispatchers.IO) {
+        register(
+            "menu-service",
+            EndpointStartupHealthCheck(httpClient) { it.get("${config.menu.baseUrl}/healthz/readiness").status.isSuccess() })
+        register(
+            "basket-service",
+            EndpointStartupHealthCheck(httpClient) { it.get("${config.basket.baseUrl}/healthz/readiness").status.isSuccess() })
+        // TODO: add start-up check for redis distributed web sessions Needs custom impl.
+    }
+
     val readinessCheck = HealthCheckRegistry(Dispatchers.IO) {
         register("menu-service", EndpointHealthCheck { it.get("${config.menu.baseUrl}/healthz/readiness") })
         register("basket-service", EndpointHealthCheck { it.get("${config.basket.baseUrl}/healthz/readiness") })
@@ -57,6 +70,7 @@
         menuService = menuService,
         basketService = basketService,
         httpClient = httpClient,
+        startupCheck = startupCheck,
         readinessCheck = readinessCheck,
         sessionStorage = RedisSessionStorage(connection.coroutines())
     )
